### ISSUE TITLE
RFC / PR Plan: Add DeepSpeed engine_impl to VERL (refactor-aware, incremental)

### ISSUE THREAD

**Issue by NenoL2001** [volcengine/verl#3361]
### Feature request

Summary
I’ve completed an initial DeepSpeed-backed engine_impl for VERL and would like to upstream it as an incremental PR that stays compatible with the ongoing refactor. This issue proposes the scope, design surface, and review plan so we can align before I open the PR.
Contact: [qling1007@yahoo.com](mailto:qling1007@yahoo.com)
Happy to adapt to your refactor plans—thanks for the guidance!

### Motivation

Motivation & Goals
Provide a first-class DeepSpeed engine option alongside existing engines (e.g., FSDP) to support ZeRO sharding and memory-efficiency features.
Keep the integration thin and modular so it survives the current codebase refactor without churn.
Unblock PPO/SFT pipelines that benefit from ZeRO-2/3, activation checkpointing, gradient accumulation, and mixed precision.
Scope of the Initial PR (MVP)
New module: engine_impl/deepspeed/ with a concrete DeepSpeedEngine that conforms to the current engine interface (init, train/eval steps, checkpoint I/O, collective ops).
Config plumbing to select engine.impl=deepspeed and pass common DeepSpeed knobs:
ZeRO stage (2/3), offload (cpu/nvme/none), gradient accumulation, activation checkpointing, BF16/FP16, grad clipping.
PPO + SFT training support (verified single-node, multi-GPU); eval path parity with existing engines.
Checkpoint save/load compatible with VERL’s checkpointer contract (weights/optimizer/scheduler state).
Minimal docs: example YAML and launcher notes.
Out of scope for the first PR (can follow up):
Pipeline parallelism, sequence parallelism.
Fully fleshed multi-node Slurm launch utilities.
Cross-engine checkpoint conversion tooling.

### Your contribution

Design Overview
Engine contract: Implements the same public methods used by trainers (init, train_step, eval_step, save_checkpoint, load_checkpoint, barriers/collectives). Avoids trainer-specific assumptions.
Optim/Shard: Uses ZeRO optimizer state partitioning; respects existing optimizer factory hooks so schedulers/optimizers remain pluggable.
Collectives: Defers to VERL’s comm utilities where available; otherwise uses torch.distributed to keep behavior consistent across engines.
Precision & memory: Supports BF16/FP16 mixed precision, activation checkpointing, gradient accumulation.
Launch: Works with torchrun and DeepSpeed launcher; honors MASTER_ADDR/MASTER_PORT/WORLD_SIZE/RANK envs.
Example Config (YAML)
engine:
  impl: deepspeed
  deepspeed:
    zero_stage: 2        # 2 or 3
    offload: none        # none | cpu | nvme
    bf16: true
    fp16: false
    gradient_accumulation_steps: 8
    activation_checkpointing: true
    grad_clip_norm: 1.0
Compatibility with the Ongoing Refactor
To reduce rework, I kept the engine surface minimal. It would help to confirm/stabilize these boundaries:
Engine base interface (method names & expected returns for train/eval steps).
DataProto / batch keys consumed by trainers.
Checkpointer contract (file layout, key names, partial-load semantics).
Metrics hooks (loss dicts, scalar reductions).
Trainer-engine wiring (no engine-specific assumptions in trainers).
I’m happy to adjust to any imminent changes here if you can point me to the target interfaces.
Status & Preliminary Results
Verified PPO and SFT single-node, multi-GPU runs with ZeRO-2 and BF16.
Sanity checks on loss parity vs. existing engine for small configs.
Checkpoint save/load round-trip tested.
(If desired, I can include a small metrics table in the PR description once we settle scope.)

Test Plan
Unit tests for engine init, optimizer wrapping, AMP settings, and checkpoint round-trip.
Trainer-level smoke tests for PPO and SFT (short runs) to confirm loss decreases and no NaNs.
CI matrix proposal (can be trimmed initially):
PyTorch 2.7–2.8, CUDA 12.x
ZeRO-2 and ZeRO-3
BF16 on Ampere+
Open Questions for Maintainers
Which engine interface is expected to be stable post-refactor? Any WIP branch I should align to?
Preference for how DeepSpeed JSON vs. YAML should be surfaced (inline YAML only, or allow external DS config file)?
Any existing helper utilities I should reuse for activation checkpointing and grad-accum hooks?
Preferred checkpoint naming/layout for engine-specific shards (especially ZeRO-3)?
Next Steps
I’ll open an incremental PR with the MVP scope above, add docs/examples, and iterate based on your feedback.
Follow-ups can bring multi-node launch helpers, pipeline/sequence parallel, and additional benchmarks.
Checklist
 I’ve read the contributing guidelines.
 Engine is modular and avoids trainer coupling.
 Basic docs and example YAML included.
 Tests for init/AMP/checkpoint; PPO/SFT smoke tests.

**Comment by glowwormX** [comment]
Looking forward to supporting deepspeed, compared to using trl, I feel that using FSDP with verl consumes more memory

**Comment by vermouth1992** [comment]
Thanks for your contribution. We are currently drafting a model engine design that is backend agnostic. This will make integration of various training backend easy and testable. We will provide detailed documentations on how to integrate. We are in the final stage of testing and it should be finished in a few weeks

**Comment by NenoL2001** [comment]
> Thanks for your contribution. We are currently drafting a model engine design that is backend agnostic. This will make integration of various training backend easy and testable. We will provide detailed documentations on how to integrate. We are in the final stage of testing and it should be finished in a few weeks

How could I start the process for a pull request?

**Comment by Kirrito-k423** [comment]
English:
This issue is being closed under the scheduled stale-issue cleanup rule.

Evidence checked:
- Issue: https://github.com/verl-project/verl/issues/3361
- Comments API: https://api.github.com/repos/verl-project/verl/issues/3361/comments
- Timeline API: https://api.github.com/repos/verl-project/verl/issues/3361/timeline
- Latest public reply or public reference from an account other than the current scheduled task account (`Kirrito-k423`): 2025-09-07T20:20:31Z
- Current run date: 2026-07-18; that latest non-self public activity is more than 6 months old.

Because the most recent non-self public reply/reference is older than 6 months, this issue meets the scheduled cleanup standard and is being closed. If the author or community can still provide a current reproduction, new evidence, or additional requirement context, please comment with those details and ask maintainers to reopen, or open a new issue with updated information.

中文：
本 issue 按定时清理规则关闭。

已检查的证据：
- Issue 页面：https://github.com/verl-project/verl/issues/3361
- Comments API：https://api.github.com/repos/verl-project/verl/issues/3361/comments
- Timeline API：https://api.github.com/repos/verl-project/verl/issues/3361/timeline
- 最近一次由当前定时任务账号（`Kirrito-k423`）之外账号产生的公开回复或公开引用时间：2025-09-07T20:20:31Z
- 本轮运行日期：2026-07-18；该最近非本人公开互动已经超过 6 个月。

由于最近一次非本人公开回复/引用已经超过 6 个月，本 issue 命中定时清理标准，因此关闭。如果作者或社区成员仍能提供当前版本的复现、新证据或更完整的需求背景，请继续评论并请求维护者重开，或用更新后的信息创建新 issue。

This response was generated by a scheduled Codex GPT5.5 issue processing task.
